pkg/route/apiserver/admission/requiredrouteannotations/admission.go

```diff
@@ -0,0 +1,352 @@
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/admission/initializer"
+	"k8s.io/client-go/informers"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	configv1 "github.com/openshift/api/config/v1"
+	grouproute "github.com/openshift/api/route"
+	configinformers "github.com/openshift/client-go/config/informers/externalversions"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	routeinformers "github.com/openshift/client-go/route/informers/externalversions"
+	routev1listers "github.com/openshift/client-go/route/listers/route/v1"
+	openshiftapiserveradmission "github.com/openshift/openshift-apiserver/pkg/admission"
+	routeapi "github.com/openshift/openshift-apiserver/pkg/route/apis/route"
+)
+
+const (
+	pluginName = "route.openshift.io/RequiredRouteAnnotations"
+	// To cover scenarios with 10,000 routes, need to wait up to 30 seconds for caches to sync
+	timeToWaitForCacheSync = 30 * time.Second
+	hstsAnnotation         = "haproxy.router.openshift.io/hsts_header"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(pluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return NewRequiredRouteAnnotations(), nil
+		})
+}
+
+type requiredRouteAnnotations struct {
+	*admission.Handler
+	routeLister   routev1listers.RouteLister
+	nsLister      corev1listers.NamespaceLister
+	ingressLister configv1listers.IngressLister
+	cachesSynced  bool
+	cachesToSync  []cache.InformerSynced
+}
+
+// Ensure that the required OpenShift admission interfaces are implemented.
+var _ = initializer.WantsExternalKubeInformerFactory(&requiredRouteAnnotations{})
+var _ = admission.ValidationInterface(&requiredRouteAnnotations{})
+var _ = openshiftapiserveradmission.WantsOpenShiftConfigInformers(&requiredRouteAnnotations{})
+var _ = openshiftapiserveradmission.WantsOpenShiftRouteInformers(&requiredRouteAnnotations{})
+
+var maxAgeRegExp = regexp.MustCompile(`max-age=(\d+)`)
+
+// Validate ensures that routes specify required annotations, and returns nil if valid.
+func (o *requiredRouteAnnotations) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	if a.GetResource().GroupResource() != grouproute.Resource("routes") {
+		return nil
+	}
+	newObject, isRoute := a.GetObject().(*routeapi.Route)
+	if !isRoute {
+		return nil
+	}
+
+	// Skip the validation if we're not updating/creating
+	switch a.GetOperation() {
+	case admission.Update, admission.Create:
+	default:
+		return nil
+	}
+
+	// Determine if there are HSTS changes in this update
+	if a.GetOperation() == admission.Update {
+		wants, has := false, false
+		var oldHSTS, newHSTS string
+
+		newMetaData, err := meta.Accessor(newObject)
+		if err != nil {
+			return err
+		}
+		newAnnotations := newMetaData.GetAnnotations()
+		newHSTS, wants = newAnnotations[hstsAnnotation]
+
+		oldObject := a.GetOldObject()
```

Comment:
Nit: consider using type assertion here to simplify the code below.
```suggestion
		oldObject := a.GetOldObject().(*routeapi.Route)
```

Reply:
done, thanks